File: python-sim/tests/test_frame.py
import pytest

from hftext.codec import pack_symbols_to_bytes
from hftext.crc16 import crc16_ccitt_false
from hftext.frame import (
    PREAMBLE_BITS,
    SYNC_BYTES,
    SYNC_BITS,
    bits_to_bytes,
    build_frame,
    build_frame_bytes,
    build_transmission,
    bytes_to_bits,
    find_sync,
    parse_frame,
    parse_frame_bytes,
    parse_frame_from_stream,
    payload_byte_count,
)


def test_build_and_parse_frame_round_trip_text():
    bits = build_frame("pu5lrk Teste")
    result = parse_frame(bits)

    assert result.frame_detected
    assert result.crc_ok
    assert result.payload_valid
    assert result.text == "pu5lrk Teste"
    assert result.length == 12


def test_build_transmission_prefixes_preamble_before_frame():
    frame_bits = build_frame("abc")
    tx_bits = build_transmission("abc")

    assert tx_bits[: len(PREAMBLE_BITS)] == PREAMBLE_BITS
    assert tx_bits[len(PREAMBLE_BITS) :] == frame_bits


def test_parse_frame_from_stream_finds_sync_after_preamble_and_noise_bits():
    stream = [0, 0, 0, 1] + build_transmission("pu5lrk Teste") + [1, 1, 1]
    result = parse_frame_from_stream(stream)

    assert result.frame_detected
    assert result.crc_ok
    assert result.payload_valid
    assert result.text == "pu5lrk Teste"
    assert result.sync_index == 4 + len(PREAMBLE_BITS)


def test_parse_frame_from_stream_skips_false_sync_until_valid_crc():
    false_frame = bytearray(build_frame_bytes("ruido"))
    false_frame[-1] ^= 0x01
    false_bits = bytes_to_bits(bytes(false_frame))
    stream = false_bits + [1, 0, 1] + build_transmission("pu5lrk ok")

    result = parse_frame_from_stream(stream)

    assert result.frame_detected
    assert result.crc_ok
    assert result.payload_valid
    assert result.text == "pu5lrk ok"
    assert result.sync_index == len(false_bits) + 3 + len(PREAMBLE_BITS)


def test_find_sync_returns_none_when_sync_is_absent():
    assert find_sync([1, 0, 1, 0, 1, 0]) is None


def test_find_sync_rejects_invalid_bits():
    with pytest.raises(ValueError, match="invalid bit"):
        find_sync([0, 1, 2] + SYNC_BITS)


def test_parse_frame_from_stream_reports_short_frame_after_sync():
    result = parse_frame_from_stream(SYNC_BITS)

    assert result.frame_detected
    assert not result.crc_ok
    assert result.error == "stream ended before length"


def test_build_frame_bytes_uses_protocol_layout():
    frame = build_frame_bytes("a")
    payload = pack_symbols_to_bytes([1])
    expected_crc = crc16_ccitt_false(payload).to_bytes(2, "big")

    assert frame == SYNC_BYTES + bytes([1]) + payload + expected_crc


def test_empty_payload_frame_is_valid():
    result = parse_frame(build_frame(""))

    assert result.frame_detected
    assert result.crc_ok
    assert result.payload_valid
    assert result.text == ""
    assert result.length == 0


def test_parse_rejects_bad_sync():
    frame = bytearray(build_frame_bytes("abc"))
    frame[0] ^= 0xFF

    result = parse_frame_bytes(bytes(frame))

    assert not result.frame_detected
    assert result.error == "sync not found"


def test_parse_rejects_length_with_bit_7_set():
    frame = bytearray(build_frame_bytes("abc"))
    frame[2] = 0x80

    result = parse_frame_bytes(bytes(frame))

    assert result.frame_detected
    assert not result.crc_ok
    assert result.error == "length bit 7 set"


def test_parse_rejects_frame_size_mismatch():
    frame = build_frame_bytes("abc")[:-1]

    result = parse_frame_bytes(frame)

    assert result.frame_detected
    assert not result.crc_ok
    assert "frame size mismatch" in result.error


def test_parse_rejects_bad_crc_without_presenting_text():
    frame = bytearray(build_frame_bytes("abc"))
    frame[-1] ^= 0x01

    result = parse_frame_bytes(bytes(frame))

    assert result.frame_detected
    assert not result.crc_ok
    assert not result.payload_valid
    assert result.text == ""
    assert result.error == "crc mismatch"


def test_parse_accepts_portuguese_modifier_symbols():
    result = parse_frame(build_frame("olá atenção"))

    assert result.frame_detected
    assert result.crc_ok
    assert result.payload_valid
    assert result.text == "olá atenção"


def test_bits_and_bytes_round_trip_msb_first():
    data = bytes([0x2D, 0xD4])

    assert bytes_to_bits(data)[:8] == [0, 0, 1, 0, 1, 1, 0, 1]
    assert bits_to_bytes(bytes_to_bits(data)) == data


def test_bits_to_bytes_rejects_invalid_bit_count():
    with pytest.raises(ValueError, match="multiple of 8"):
        bits_to_bytes([1, 0, 1])


def test_bits_to_bytes_rejects_invalid_bit_value():
    with pytest.raises(ValueError, match="invalid bit"):
        bits_to_bytes([0, 0, 0, 0, 0, 0, 0, 2])


def test_payload_byte_count_matches_six_bit_packing():
    assert payload_byte_count(0) == 0
    assert payload_byte_count(1) == 1
    assert payload_byte_count(4) == 3
    assert payload_byte_count(127) == 96

File: core/tests/test_frame.cpp
#include "hftext_crc16.h"
#include "hftext_frame.h"

#include <cassert>
#include <cstdint>
#include <stdexcept>
#include <string>
#include <vector>

namespace {

void assertValidText(const hftext::DecodeResult& result, const std::string& expectedText, int expectedLength) {
    assert(result.frameDetected);
    assert(result.crcOk);
    assert(result.payloadValid);
    assert(result.text == expectedText);
    assert(result.length == expectedLength);
}

bool contains(const std::string& text, const std::string& needle) {
    return text.find(needle) != std::string::npos;
}

}  // namespace

int main() {
    auto result = hftext::parseFrame(hftext::buildFrame("pu5lrk Teste"));
    assertValidText(result, "pu5lrk Teste", 13);

    const auto frameBits = hftext::buildFrame("abc");
    const auto txBits = hftext::buildTransmission("abc");
    const auto preamble = hftext::defaultPreambleBits();
    assert(std::vector<std::uint8_t>(txBits.begin(), txBits.begin() + preamble.size()) == preamble);
    assert(std::vector<std::uint8_t>(txBits.begin() + preamble.size(), txBits.end()) == frameBits);

    std::vector<std::uint8_t> stream = {0, 0, 0, 1};
    const auto transmission = hftext::buildTransmission("pu5lrk Teste");
    stream.insert(stream.end(), transmission.begin(), transmission.end());
    stream.insert(stream.end(), {1, 1, 1});
    result = hftext::parseFrameFromStream(stream);
    assertValidText(result, "pu5lrk Teste", 13);
    assert(result.syncIndex == static_cast<int>(4 + preamble.size()));

    assert(hftext::findSync({1, 0, 1, 0, 1, 0}) == -1);

    bool invalidBitRejected = false;
    try {
        auto badBits = std::vector<std::uint8_t>{0, 1, 2};
        const auto sync = hftext::syncBits();
        badBits.insert(badBits.end(), sync.begin(), sync.end());
        (void)hftext::findSync(badBits);
    } catch (const std::invalid_argument&) {
        invalidBitRejected = true;
    }
    assert(invalidBitRejected);

    result = hftext::parseFrameFromStream(hftext::syncBits());
    assert(result.frameDetected);
    assert(!result.crcOk);
    assert(result.error == "stream ended before length");

    const auto oneSymbolPayload = hftext::packSymbolsToBytes({1});
    auto expectedFrame = hftext::syncBytes();
    expectedFrame.push_back(1);
    expectedFrame.insert(expectedFrame.end(), oneSymbolPayload.begin(), oneSymbolPayload.end());
    const auto expectedCrc = hftext::crc16CcittFalse(oneSymbolPayload);
    expectedFrame.push_back(static_cast<std::uint8_t>((expectedCrc >> 8) & 0xFF));
    expectedFrame.push_back(static_cast<std::uint8_t>(expectedCrc & 0xFF));
    assert(hftext::buildFrameBytes("a") == expectedFrame);

    result = hftext::parseFrame(hftext::buildFrame(""));
    assertValidText(result, "", 0);

    auto badSync = hftext::buildFrameBytes("abc");
    badSync[0] ^= 0xFF;
    result = hftext::parseFrameBytes(badSync);
    assert(!result.frameDetected);
    assert(result.error == "sync not found");

    auto badLength = hftext::buildFrameBytes("abc");
    badLength[2] = 0x80;
    result = hftext::parseFrameBytes(badLength);
    assert(result.frameDetected);
    assert(!result.crcOk);
    assert(result.error == "length bit 7 set");

    auto shortFrame = hftext::buildFrameBytes("abc");
    shortFrame.pop_back();
    result = hftext::parseFrameBytes(shortFrame);
    assert(result.frameDetected);
    assert(!result.crcOk);
    assert(contains(result.error, "frame size mismatch"));

    auto badCrc = hftext::buildFrameBytes("abc");
    badCrc.back() ^= 0x01;
    result = hftext::parseFrameBytes(badCrc);
    assert(result.frameDetected);
    assert(!result.crcOk);
    assert(!result.payloadValid);
    assert(result.text.empty());
    assert(result.error == "crc mismatch");

    const auto reservedPayload = hftext::packSymbolsToBytes({61});
    auto reservedFrame = hftext::syncBytes();
    reservedFrame.push_back(1);
    reservedFrame.insert(reservedFrame.end(), reservedPayload.begin(), reservedPayload.end());
    const auto reservedCrc = hftext::crc16CcittFalse(reservedPayload);
    reservedFrame.push_back(static_cast<std::uint8_t>((reservedCrc >> 8) & 0xFF));
    reservedFrame.push_back(static_cast<std::uint8_t>(reservedCrc & 0xFF));
    result = hftext::parseFrameBytes(reservedFrame);
    assert(result.frameDetected);
    assert(result.crcOk);
    assert(!result.payloadValid);
    assert(result.text.empty());
    assert(contains(result.error, "invalid symbol"));

    const std::vector<std::uint8_t> syncBytes = {0x2D, 0xD4};
    const auto convertedSyncBits = hftext::bytesToBits(syncBytes);
    assert(std::vector<std::uint8_t>(convertedSyncBits.begin(), convertedSyncBits.begin() + 8)
        == std::vector<std::uint8_t>({0, 0, 1, 0, 1, 1, 0, 1}));
    assert(hftext::bitsToBytes(convertedSyncBits) == syncBytes);

    bool bitCountRejected = false;
    try {
        (void)hftext::bitsToBytes({1, 0, 1});
    } catch (const std::invalid_argument&) {
        bitCountRejected = true;
    }
    assert(bitCountRejected);

    bool bitValueRejected = false;
    try {
        (void)hftext::bitsToBytes({0, 0, 0, 0, 0, 0, 0, 2});
    } catch (const std::invalid_argument&) {
        bitValueRejected = true;
    }
    assert(bitValueRejected);

    assert(hftext::payloadByteCount(0) == 0);
    assert(hftext::payloadByteCount(1) == 1);
    assert(hftext::payloadByteCount(4) == 3);
    assert(hftext::payloadByteCount(127) == 96);

    assert(hftext::packSymbolsToBytes({1, 2, 3, 0}) == std::vector<std::uint8_t>({0x04, 0x20, 0xC0}));
    assert(hftext::unpackSymbolsFromBytes({0x04}, 1) == std::vector<std::uint8_t>({1}));
}
